### PR TITLE
feat: support metadata table "snapshots"

### PR DESCRIPTION
**Pull request by xxchan**
#823

In Iceberg Spark/Flink there are metadata tables that provide information around the table: https://iceberg.apache.org/docs/latest/spark-queries/#inspecting-tables

Supporting this in `iceberg-rust` allows other engines (like RisingWave) to support these "standard" metadata tables.

In this PR, I added a trait for metadata tables and the first implementation for the `snapshots` metadata table. If it looks good, we can add more later.

reference implementation: 
    - <https://py.iceberg.apache.org/api/#inspecting-tables>
    - https://github.com/apache/iceberg-python/blob/main/pyiceberg/table/inspect.py#L58-L90

### CODE REVIEW

File: crates/iceberg/src/metadata_scan.rs
Comment (xxchan) [inline crates/iceberg/src/metadata_scan.rs:57]: I need to make this separate and public, so engines can get the schema first without fetching the data.

File: crates/iceberg/src/scan.rs
Comment (xxchan) [inline crates/iceberg/src/scan.rs:986]: Make this `pub` so that we can share it in the `metadata_scan` test.
We can move it to other places like `test_utils`, but I want to avoid unnecessary changes in the PR.

**Comment by xxchan** [comment]
cc @liurenjie1024 @Xuanwo @Fokko @sdd

File: crates/iceberg/src/metadata_scan.rs
Comment (rshkv) [inline crates/iceberg/src/metadata_scan.rs:39]: Wondering what you think about

```suggestion
pub struct MetadataScan<'a> {
    table: &'a Table,
}
```

so we can get at `FileIO`.

File: crates/iceberg/src/metadata_scan.rs
Comment (rshkv) [inline crates/iceberg/src/metadata_scan.rs:68]: Thoughts on making async because some of the metadata scans will do async reads of metadata files.

File: crates/iceberg/src/metadata_scan.rs
Comment (rshkv) [reply crates/iceberg/src/metadata_scan.rs:65]: Some of the metadata tables (e.g. `entires`) have schemas that depend on the table schema. E.g. `readable_metrics` is a struct with field per column. Might be worth having this take a `scan` as well or do so in another PR.

**Comment by xxchan** [comment]
@rshkv Thanks for the review and the follow-up work. I just want to wait for the maintainers to make sure the framework looks good before applying any suggestions :p  

BTW I think we can also change the details like signatures and fields on demand later.

File: crates/iceberg/src/metadata_scan.rs
Comment (flaneur2020) [reply crates/iceberg/src/metadata_scan.rs:68]: +1 for this, the manifests table requires an async call `load_manifest_list()` with a dependency on `FileIO`.

**Comment by rshkv** [comment]
Yeah, totally! 

> BTW I think we can also change the details like signatures and fields on demand later.

Yeah that's what I was worried about. Sg. 👍🏻

File: crates/iceberg/src/metadata_scan.rs
Comment (Xuanwo) [inline crates/iceberg/src/metadata_scan.rs:63]: Hi, I'm a bit confused about why we need this trait. Could you show how we would use it? Is there a use case that we need `impl MetadataTable` or `dyn MetadataTable`? Does it make sense to provide API like `table.snapshot_table()`?

Comment (xxchan) [reply crates/iceberg/src/metadata_scan.rs:63]: Yes, it's not a must-have. Just an common interface for different metadata tables.

One practical motivation is that I want to ensure `schema` method exists, like `SnapshotsTable::schema()`.

Comment (Xuanwo) [reply crates/iceberg/src/metadata_scan.rs:63]: If so, directly constructing structs makes more sense to me, as it makes the API more intuitive to use.

Would you like to give it a try?

Comment (xxchan) [reply crates/iceberg/src/metadata_scan.rs:63]: Previously I thought the schema is always a constant, so I designed the interface so that query engine can get the schema without loading the table from catalog. 

But it seems not the case. As @rshkv mentioned https://github.com/apache/iceberg-rust/pull/822#discussion_r1898945557

e.g., `readable_metrics` is dynamic https://py.iceberg.apache.org/api/#entries


Therefore, what about this alternative design:
```rust
// user API
table.metadata_scan().snapshots().schema()
table.metadata_scan().snapshots().scan()

// impl

impl MetadataScan {
    /// Returns the snapshots of the table.
    pub fn snapshots(&self) -> Result<RecordBatch> {
        SnapshotsTable{ metadata: self }
    }
}

struct SnapshotsTable<'a> {
    metadata: &'a MetadataScan,
}

// trait is just to ensure interface, and add doc
pub trait MetadataTable {
    /// Returns the schema of the metadata table.
    fn schema(&self) -> Schema;

    /// Scans the metadata table.
    fn scan(&self) -> Result<RecordBatch>;
}
```

Comment (xxchan) [reply crates/iceberg/src/metadata_scan.rs:63]: Oh, didn't see you comment above. I think whether or not to have the trait are both OK to me

Comment (xxchan) [reply crates/iceberg/src/metadata_scan.rs]: updated

Comment (flaneur2020) [reply crates/iceberg/src/metadata_scan.rs]: :+1: removing this trait definition looks good to me

File: crates/iceberg/src/metadata_scan.rs
Comment (Xuanwo) [inline crates/iceberg/src/metadata_scan.rs:40]: `MetadataScan` itself appears to be a plan, so it's somewhat unusual to encounter code like:

```rust
table.metadata_scan().snapshots().scan()
```

How about using `MetadataTable` instead?

```rust
table.metadata_table().snapshots().scan()
```

And `MetadataTable` can just serve a new type of `Table` will provides a category for metadata related APIs.

```rust
pub struct MetadataTable(Table);
```

File: crates/iceberg/src/metadata_scan.rs
Comment (Xuanwo) [inline crates/iceberg/src/metadata_scan.rs:46]: We can expect users to obtain the struct through `Table` instead; perhaps we don't need those public APIs. How about keeping it `pub(crate)`?

**Review by Xuanwo — COMMENTED** [review]
Hi, it looks much better now. Only some minor questions.

**Comment by xxchan** [comment]
Thanks for the review. Updated. Please take a another look.

File: crates/iceberg/src/lib.rs
Comment (Xuanwo) [inline crates/iceberg/src/lib.rs:76]: Should be changed in to `metadata_table`.

File: crates/iceberg/src/table.rs
Comment (Xuanwo) [inline crates/iceberg/src/table.rs:206]: Better to use `&self` here to align with other APIs.

**Review by Xuanwo — APPROVED** [review]
This PR is already in good shape. Let's merge it to unblock other pending PRs. Suggested changes can be made in a separate PR.

**Pull request merged**
